### PR TITLE
issue: 977994 Fixed connection timed out for TCP fast open/close sockets

### PR DESCRIPTION
**Pull request by serglelv**
Fixed infinite epoll_wait for EAGAIN and EBUSY states and fixed timed out termination. Please review.
 @OphirMunk @igor-ivanov 

Signed-off-by: sergeyly <sergeyly@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1219/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/860/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
@serglelv please fix Jenkins

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1229/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/866/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1230/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/867/ for details (Mellanox internal link).

**Comment by serglelv** [comment]
Fix was disabled and Jenkins passed. Investigation is continued.
@OphirMunk @igor-ivanov

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1232/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1253/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/874/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1254/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/875/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1255/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1256/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1262/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1265/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/876/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1271/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1272/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1278/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1283/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1290/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1291/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1296/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_tcp.cpp [hunk 1]
```diff
@@ -2927,21 +2928,41 @@ err_t sockinfo_tcp::connect_lwip_cb(void *arg, struct tcp_pcb *tpcb, err_t err)
...
+		if (m_loops_timer.time_left_msec() == -1) {
+			m_n_wait_for_conn_ready_threshold++;
+		}
+
 		if (rx_wait(poll_count, m_b_blocking) < 0) {
 			si_tcp_logdbg("connect interrupted");
 			return -1;
 		}
+
+		if (errno == EAGAIN) {
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: An errno is set after returning from an API. The API in this while() loop is rx_wait().
However if rx_wait()<0 and sets errno we return with -1 and do not reach this line. So what is the meaning of testing errno at this point?

File: src/vma/sock/sockinfo_tcp.cpp [hunk 2]
```diff
@@ -2927,21 +2928,41 @@ err_t sockinfo_tcp::connect_lwip_cb(void *arg, struct tcp_pcb *tpcb, err_t err)
...
+			if (m_n_wait_for_conn_ready_threshold > WAIT_CONN_EAGAIN_THRESHOLD) {
+				m_loops_timer.set_timeout_msec(10);
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: @serglelv By setting m_loops_timer.set_timeout_msec - you are changing a value related to SO_RCVTIMEO.
A user who will run getsockopt(...,SO_RCVTIMEO,...) will get a value he had never set in setsockopt(...SO_RCVTIMEO,...)
Please explain

File: src/vma/sock/sockinfo_tcp.cpp [hunk 3]
```diff
@@ -2927,21 +2928,41 @@ err_t sockinfo_tcp::connect_lwip_cb(void *arg, struct tcp_pcb *tpcb, err_t err)
...
+			}
+		}
+		if (errno == EBUSY) {
+			if (m_n_wait_for_conn_ready_threshold > WAIT_CONN_EBUSY_THRESHOLD) {
+				m_loops_timer.set_timeout_msec(10);
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: @serglelv Why do you keep setting set_timeout_msec(10)? Does it have a new effect after the first time it was set?

File: src/vma/sock/sockinfo_tcp.cpp [hunk 4]
```diff
@@ -2927,21 +2928,41 @@ err_t sockinfo_tcp::connect_lwip_cb(void *arg, struct tcp_pcb *tpcb, err_t err)
 	return ERR_OK;
 }
 
+#define	WAIT_CONN_EAGAIN_THRESHOLD	200
```
Comment (serglelv) [inline src/vma/sock/sockinfo_tcp.cpp]: The fix will be aligned with customer scenario.

File: src/vma/sock/sockinfo_tcp.cpp (hunk 1 above)
Comment (serglelv) [inline src/vma/sock/sockinfo_tcp.cpp]: The fix will be aligned with customer scenario.

File: src/vma/sock/sockinfo_tcp.cpp (hunk 2 above)
Comment (serglelv) [inline src/vma/sock/sockinfo_tcp.cpp]: Yes, you are right. We don't need to changing m_loops_timer.set_timeout_msec It should be changed.

File: src/vma/sock/sockinfo_tcp.cpp (hunk 3 above)
Comment (serglelv) [inline src/vma/sock/sockinfo_tcp.cpp]: Yes, it does. Will be changed as well.

**Comment by OphirMunk** [comment]
Closed due to new root cause suggested by @serglelv 

The root cause of the issue is the value of MSL (Maximum Segment Lifetime).
Currently MSL is 60 sec.
tcp_impl.h L. 130 
#define TCP_MSL 60000UL /* The maximum segment lifetime in milliseconds */
With the current MSL ~ 100 connections are active in TIME_WAIT state and eventually the new connection cannot be created and test failed.
The test tcpOpenCloseClient.py/tcpOpenCloseserver.py (without any modifications) was successfully passed for 
MSL 1000 milliseconds i.e. #define TCP_MSL 1000UL /* The maximum segment lifetime in milliseconds */
It is interesting but without VMA (Linux sockets) the test was passed and MSL was ~800 milliseconds.
TIME-WAIT 0 0 12.12.12.13:5005 12.12.12.12:57258 timer:(timewait,735ms,0) 
BTW, in Linux MSL is configurable.

It seems the MSL parameter should be configurable in VMA.

**Pull request closed without merging**